posts/inside-rust/2020-11-11-source-based-code-coverage.md

```diff
@@ -0,0 +1,79 @@
+---
+layout: post
+title: Source-based code coverage in nightly
+author: Tyler Mandry
+team: The Compiler Team <https://www.rust-lang.org/governance/teams/compiler>
+---
+
+
+Support has landed in the nightly compiler for source-based code coverage,
+and we want your help testing it!
+
+
+# What is _source-based_ code coverage, exactly?
+
+You may already be familiar with code coverage, which shows you which lines
+of code execute. Code coverage is usually applied to tests to find out which
+code is actually being tested and which code isn’t.
+
+Nightly Rust already supports another kind of source code coverage, commonly
+called gcov, which relies on debug info to map from LLVM IR to lines of
+source code. Instrumentation is then added in the LLVM backend during code
+generation to count how many times each line is run.
+
+However, since LLVM doesn’t know exactly how Rust code is structured, there’s
+a lot lost in the translation between Rust source and LLVM IR. Line-level
+granularity is sometimes too coarse, and debug info can be unreliable,
+especially when building in release mode. The result is coverage reports that
+only show an approximation of what code actually executed.
+
+Source-based code coverage instrumentation happens directly in the Rust
+compiler. The compiler knows how to handle things down to individual
+branches, even within the same line. That means that things like
+short-circuited conditionals, closures, and match guards are all precisely
+counted. Counters are injected as additional code statements, so the compiler
```

Comment:
It might be good to mention this happens at the MIR level in contrast with the gcov support which happens during codegen per the previous section.